src/webgpu/api/operation/rendering/blending.spec.ts

```diff
@@ -7,7 +8,269 @@ TODO:
+      return mapColor(src, (v, k) => v + dst[k]);
+    case 'max':
+      return mapColor(src, (v, k) => Math.max(v, dst[k]));
+    case 'min':
+      return mapColor(src, (v, k) => Math.min(v, dst[k]));
+    case 'reverse-subtract':
+      return mapColor(src, (v, k) => dst[k] - v);
+    case 'subtract':
+      return mapColor(src, (v, k) => v - dst[k]);
+  }
+}
+
+g.test('GPUBlendComponent')
+  .desc(
+    `Test all combinations of parameters for GPUBlendComponent.
+
+  Tests that parameters are correctly passed to the backend API and blend computations
+  are done correctly by blending a single pixel. The test uses rgba32float as the format
+  to avoid checking clamping behavior (tested in api,operation,rendering,blending:clamp,*).
+
+  Params:
+    - component= {color, alpha} - whether to test blending the color or the alpha component.
+    - srcFactor= {...all GPUBlendFactors}
+    - dstFactor= {...all GPUBlendFactors}
+    - operation= {...all GPUBlendOperations}
+  `)
+  .cases(
+    params() //
+      .combine(poptions('component', ['color', 'alpha'] as const))
+      .combine(poptions('srcFactor', kBlendFactors))
+      .combine(poptions('dstFactor', kBlendFactors))
+      .combine(poptions('operation', kBlendOperations))
+  )
+  .fn(t => {
+    const textureFormat: GPUTextureFormat = 'rgba32float';
+    const srcColor: GPUColorDict = { r: 0.11, g: 0.61, b: 0.81, a: 0.44 };
+    const dstColor: GPUColorDict = { r: 0.51, g: 0.22, b: 0.71, a: 0.33 };
+    const varColor: GPUColorDict = { r: 0.91, g: 0.82, b: 0.73, a: 0.64 };
+
+    const srcFactor = computeBlendFactor(srcColor, dstColor, varColor, t.params.srcFactor);
+    const dstFactor = computeBlendFactor(srcColor, dstColor, varColor, t.params.dstFactor);
+
+    const expectedColor =
+      t.params.operation === 'min' || t.params.operation === 'max'
+        ? computeBlendOperation(
+          srcColor, dstColor,
+          t.params.operation
+        ) : computeBlendOperation(
+          mapColor(srcColor, (v, k) => srcFactor[k] * v),
+          mapColor(dstColor, (v, k) => dstFactor[k] * v),
+          t.params.operation
+        );
+
+    switch (t.params.component) {
+      case 'color':
+        expectedColor.a = srcColor.a;
+        break;
+      case 'alpha':
+        expectedColor.r = srcColor.r;
+        expectedColor.g = srcColor.g;
+        expectedColor.b = srcColor.b;
+        break;
+    }
+
+    const pipeline = t.device.createRenderPipeline({
+      fragment: {
+        targets: [
+          {
+            format: textureFormat,
+            blend: {
+              // Set both color/alpha to not do blending.
+              color: {
+                srcFactor: 'one',
+                dstFactor: 'zero',
+                operation: 'add',
+              },
+              alpha: {
+                srcFactor: 'one',
+                dstFactor: 'zero',
+                operation: 'add',
+              },
+              // And then override the component we're testing.
```

Comment:
:galaxy_brain:

But also the above values for `color` and `alpha` are the defaults so you should be able to just omit them for the same effect

Reply:
nvm I was wrong about this, the dictionaries have defaults but `color` and `alpha` are required.